### PR TITLE
Allow NSGs to be added onto NICs at creation and update

### PR DESCRIPTION
**Pull request by marc-sensenich**
Signed-off-by: Marc Sensenich <sensenichm91@gmail.com>

##### SUMMARY

When creating a NIC and associating an existing NSG at creation, it does not associate the NSG to the NIC. There is also the case of disassociating of the NSG from NICs when updating them if there was a NSG already associated to a NIC.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

azure_rm_networkinterface
##### ANSIBLE VERSION

```paste below
ansible 2.6.3
```

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: lib/ansible/modules/cloud/azure/azure_rm_networkinterface.py [hunk 1]
```diff
@@ -528,22 +546,24 @@ def exec_module(self, **kwargs):
...
+                update_tags, results['tags'] = self.update_tags(
+                    results['tags'])
                 if update_tags:
                     changed = True
 
-                if self.create_with_security_group != bool(results.get('network_security_group')):
-                    self.log("CHANGED: add or remove network interface {0} network security group".format(self.name))
+                if self.create_with_security_group and not bool(results.get('network_security_group')):
+                    self.log(
+                        "CHANGED: add or remove network interface {0} network security group".format(self.name))
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_networkinterface.py]: Here only detect adding change

**Comment by Fred-sun** [comment]
@marc-sensenich Thanks for the contribution, last commit unsuccessful, Could you help recheck this PR and push to forward? Thanks!

File: lib/ansible/modules/cloud/azure/azure_rm_networkinterface.py [hunk 2]
```diff
@@ -559,50 +579,68 @@ def exec_module(self, **kwargs):
...
                     self.log("CHANGED: DNS servers set to {0} (previously {1})".format(
                         ", ".join(_dns_servers_set),
                         ", ".join(_dns_servers_res)))
                     changed = True
 
-                if not changed:
-                    nsg = self.get_security_group(self.security_group['resource_group'], self.security_group['name'])
-                    if nsg and results.get('network_security_group') and results['network_security_group'].get('id') != nsg.id:
-                        self.log("CHANGED: network interface {0} network security group".format(self.name))
+                if nsg:
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_networkinterface.py]: It is not cover the remove nsg

**Review by yuwzho — CHANGES_REQUESTED** [review]

**Comment by yuwzho** [comment]
Hi @marc-sensenich Could you please drive more details of the bug? For the "associate an existing nsg to nic", I believe it is fixed in #39213 .

**Comment by mattclay** [comment]
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/85173/79/tests

**Comment by marc-sensenich** [comment]
@mattclay @Fred-sun addressed the initial issues on the broken tests with `7fdc1ae`

@yuwzho The issue that is occurring in my use case is that I am creating a Network Security Group and then associating the Network Security Group with the Network Interface in a later task. When running the Ansible Module as it is in `devel` if the NSG has been found and set to the variable `nsg` in the Ansible module, it will get overwritten to `None` if not `create_default_security_group` on the following line https://github.com/ansible/ansible/blame/08c392477ebc98111cee950df16af943a5726c8c/lib/ansible/modules/cloud/azure/azure_rm_networkinterface.py#L641.

File: lib/ansible/modules/cloud/azure/azure_rm_networkinterface.py (hunk 1 above)
Comment (marc-sensenich) [reply lib/ansible/modules/cloud/azure/azure_rm_networkinterface.py]: Addressing these comments now

**Comment by yuwzho** [comment]
@marc-sensenich the current logic is something like the following:

1. check whether the `self.security_group` is set, if not, use `self.name` as default, and parse the id/dict/name to a resource dict which contains subscriptionID/resourceGroup/name

   https://github.com/ansible/ansible/blob/ac9d506a617176510e39f8d69a020855527c19fa/lib/ansible/modules/cloud/azure/azure_rm_networkinterface.py#L512

2. If `self.create_with_security_group` (`True` by default), then call the `create_default_security_group`:

   https://github.com/ansible/ansible/blob/ac9d506a617176510e39f8d69a020855527c19fa/lib/ansible/modules/cloud/azure/azure_rm_networkinterface.py#L637-L641

   OTH, the `create_default_security_group` will check whether the nsg exist, return the existing one or new created one. 
    
   https://github.com/ansible/ansible/blob/ac9d506a617176510e39f8d69a020855527c19fa/lib/ansible/module_utils/azure_rm_common.py#L606-L614

So if you specify a custom nsg in the nic, this module will do nothing for the nsg but associate it to the nic.

For your case, I write this playbook running with devel branch, and it shows the nsg is successfully attach to nic without any other change

```yaml
---
- hosts: localhost
  tasks:
    - name: Prepare variables
      set_fact:
        resource_group: "ansiblensginnic"
    
    - name: Create resource group
      azure_rm_resourcegroup:
          name: "{{ resource_group }}"
          location: eastus
    
    - name: Create virtual network
      azure_rm_virtualnetwork:
          resource_group: "{{ resource_group }}"
          name: "tn000"
          address_prefixes: "10.10.0.0/16"
    
    - name: Add subnet
      azure_rm_subnet:
          resource_group: "{{ resource_group }}"
          name: "tn000"
          address_prefix: "10.10.0.0/24"
          virtual_network: "tn000"
    
    - name: create public ip
      azure_rm_publicipaddress:
        name: "pip000"
        resource_group: '{{ resource_group }}'
    
    - name: Create security group
      azure_rm_securitygroup:
          resource_group: "{{ resource_group }}"
          name: testnsg
          rules:
              - name: AllowSSH
                protocol: Tcp
                source_address_prefix: 174.109.158.0/24
                destination_port_range: 22
                access: Allow
                priority: 101
                direction: Inbound
      register: nsg
    
    - debug: 
        var: nsg.state
    
    - name: Create NIC
      azure_rm_networkinterface:
          resource_group: "{{ resource_group }}"
          name: "tn000"
          virtual_network: "tn000"
          subnet: "tn000"
          public_ip_name: "tn000"
          public_ip_allocation_method: Static
          security_group: testnsg
      register: output
    
    - debug:
        var: output.state.network_security_group
    
    - name: retrive nsg detail
      azure_rm_securitygroup_facts:
        resource_group: "{{ resource_group }}"
        name: "{{ output.state.network_security_group.name }}"
      register: nsg

    - debug:
         var: nsg.ansible_facts.azure_securitygroups[0]
    
    - name: tear down
      azure_rm_resourcegroup:
         state: absent
         name: "{{ resource_group }}"
         force: true
```

**Comment by marc-sensenich** [comment]
Thanks for your case @yuwzho. I've ran that and saw the same results, my bug seems to be around the case of using `create_with_security_group` and defining a `security_group`

With your test case it fails given the `create_with_security_group` is set to `False`
```yaml
- name: Create NIC
   azure_rm_networkinterface:
          resource_group: "{{ resource_group }}"
          name: "{{ nic_name }}"
          virtual_network: "{{ virtual_network_name }}"
          subnet: "{{ subnet_name }}"
          security_group: "{{ security_group }}"
          create_with_security_group: False
    register: output
    
    - debug:
        var: output.state.network_security_group
    
    - name: retrive nsg detail
      azure_rm_securitygroup_facts:
        resource_group: "{{ resource_group }}"
        name: "{{ output.state.network_security_group.name }}"
      register: nsg

    - debug:
         var: nsg.ansible_facts.azure_securitygroups[0]
```

**Comment by Fred-sun** [comment]
@yuwzho

**Comment by yuwzho** [comment]
@marc-sensenich What do you mean for `fail`? throw exception in the `azure_rm_networkinterface` module? For a expected behavior, if  `create_with_security_group` is set to `false`, the NIC should be associated with no NSG.

**Comment by Fred-sun** [comment]
@marc-sensenich Thanks for the contribution, Do you have any updated according by @yuwzho's comments?
Another, if you set create_with_security_group is False, Create NIC will not associate security group, So you use output to get network_security_group.name value is null, will meet error as follow describe, Thanks!
```e.g.
TASK [retrive nsg detail] ******************************************************
task path: /home/fred/task/46003/1.yml:24
fatal: [localhost]: FAILED! => {
    "msg": "The task includes an option with an undefined variable. The error was: 'None' has no attribute 'name'\n\nThe error appears to have been in '/home/fred/task/46003/1.yml': line 24, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n    - name: retrive nsg detail\n      ^ here\n"
}

**Comment by Fred-sun** [comment]
@marc-sensenich Thanks for your contribution, Are you stall here? Could you help recheck this according by yuwzho's advice. Thanks!

**Comment by marc-sensenich** [comment]
@Fred-sun I'm still here and based on the @yuwzho presented I was able to replicate the expected behavior. More of a documentation and validation issue from the module viewpoint. I think this work could be more of a parameter clarification and input enforcement PR with changing `create_with_security_group` to `create_with_default_security_group` and making the `security_group` and `create_with_security_group` parameters mutually exclusive. Thoughts?

**Comment by yuwzho** [comment]
@marc-sensenich Make sense. Considering the compatible issue, we need to keep the `create_with_security_group` as an alias

**Comment by Fred-sun** [comment]
@marc-sensenich Thanks for your contribution, Are you still tracking on this issue? Could you help confirm how about the PR going? merged or further review? Thanks!

**Comment by Fred-sun** [comment]
kindly ping

**Comment by marc-sensenich** [comment]
@Fred-sun I'm going to close this PR, I haven't had the cycles to look at this. I'll describe the behavior in an issue and leave it for the owners to grab

**Pull request closed without merging**